File: javascript/92.反转链表-ii.js
/*
 * @lc app=leetcode.cn id=92 lang=javascript
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    if (!head || !head.next) {
        return head;
    };
    var sentryNode = new ListNode(0);
    sentryNode.next = head;
    var preStartNode = sentryNode;
    // 找到要翻转的 m 的前置节点
    for (var i = 1; i < m; i++) {
        preStartNode = preStartNode.next;
    };
    var start = preStartNode.next;
    var tail = start.next;
    for (var j = m ; j <= n; j++) { // 另一种解法
        start.next = tail.next;
        tail.next = preStartNode.next;
        preStartNode.next = tail;
        tail = start.next;
    };
    return sentryNode.next;

};
// @lc code=end

File: java/92.反转链表-ii.java
/*
 * @lc app=leetcode.cn id=92 lang=java
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 * 
 * @param <ListNode>
 */
class Solution<ListNode> {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode sentryNode = new ListNode(0);
        sentryNode.next = head;
        ListNode preStartNode = sentryNode;
        // 找到要翻转的 m 的前置节点
        for (int i = 1; i < m; i++) {
            preStartNode = preStartNode.next;
        }
        // 翻转 m - n
        ListNode preNode = null; // 开始翻转的时候的前置节点
        ListNode currNode = preStartNode.next; // 开始翻转的当前节点
        for (int j = m ; j <= n; j++) {
            ListNode next = currNode.next;
            currNode.next = preNode;
            preNode = currNode;
            currNode = next; 
        }
        // 讲翻转的链表和未翻转的链表拼接
        preStartNode.next.next = currNode;
        preStartNode.next = preNode;
        return sentryNode.next;
    }
}
// @lc code=end

